### ISSUE TITLE
Use ATrace_isEnabled as a signal that timeline events should go to Android system tracing

### ISSUE THREAD

**Issue by zanderso** [dart-lang/sdk#45143]
This might obviate the need for Flutter's `--trace-systrace` flag such that timeline events would automagically go to Android system tracing when Android tooling is attached, and would otherwise be sent to Flutter/Dart's tooling otherwise.

However, I'm not sure whether the flag would need to be checked for every event, or only checked once at VM/tracing startup.

https://developer.android.com/ndk/reference/group/tracing#group___tracing_1gad839953d2c6d4189ec80f7b028f7f778

**Comment by zanderso** [comment]
From chat, I am told that the flag would need to be checked on every event, or at least periodically.

**Comment by dnfield** [comment]
What's the devexp like if you're viewing the timeline in the observatory and then turn on tracing on your device?

Would the observatory let you know that it's no longer recording and you have to view the systrace?

**Comment by zanderso** [comment]
@dnfield I suspect this might need a little investigation before getting to those questions. If checking the flag on every event is too expensive, the VM might just check once at startup, for example.

/cc @bkonyi

**Comment by dnfield** [comment]
The performance question is important, but I'm more asking about what happens if a developer is trying to view the timeline in the observatory and suddenly can't because systrace is being used.

**Comment by zanderso** [comment]
Another way to approach this would be for the VM to provide an API to switch the tracing backends at runtime, and the policy to decide when to switch could be managed by embedders.

**Comment by xster** [comment]
A beginner question, why can't the the events always be logged to the service isolate but also to conditionally also send to perfetto if ATrace_isEnabled is on?

**Comment by xster** [comment]
Another point of consideration from @jacob314, the implementation of the sending to perfetto can't be in the service isolate since it needs to work in release mode as well.

**Comment by dnfield** [comment]
Last time I asked @rmacnak-google it was concerns over affecting performance

**Comment by xster** [comment]
Ryan, which mode are you worried about? 

In release mode: it shouldn't matter too much since we can't send the trace events to the observatory/service isolate anyway
In profile mode: here the concern is that with trace on, it'll behave a bit differently than with traces off? Can we try to quantify it? What parts would come from the observatory logging vs the perfetto logging?
In debug mode: we don't care right?

**Comment by dnfield** [comment]
If we want to do both why not just do it in the embedding? @chinmaygarde

**Comment by xster** [comment]
Doing some parts of it in the embedding sounds reasonable. How would we hook out of the common Dart Timeline APIs to let the embedding pipe it to Android? How would the threading work?

**Comment by chinmaygarde** [comment]
What problem are we trying to solve here? `--trace-systrace` works fine right? IMO, its fine to explicitly specify which tracing backend you want to use instead of depending on signals from the environment.

**Comment by xster** [comment]
Hard to discuss details here. b/195083760. We don't know if we should send to systrace or to the service isolate at launch time.

**Comment by chinmaygarde** [comment]
I read the context on the linked bug. So, the service isolate is not related to logging events to the Dart timeline. In release modes, none of the Dart VM timeline code is even being compiled. So even if we made calls into `Dart_TimelineEvent` based on `ATrace_isEnabled`, those calls would be no-ops.

We could turn on these subsystems in `release` modes. But, at that point, I contend that you should be using the `profile` variant instead of making `release` mode work more like `profile`. That's why we build and vend such a variant.

**Comment by xster** [comment]
ya, it wouldn't be a trivial thing to turn on. I imagine we'd have to split the rest of the profile features, service protocol, observatory etc etc from this so that it's present in release mode and can log to systrace. Android itself also doesn't disable calls into its `Trace` API when in release mode.

**Comment by zanderso** [comment]
> ya, it wouldn't be a trivial thing to turn on. I imagine we'd have to split the rest of the profile features, service protocol, observatory etc etc from this so that it's present in release mode and can log to systrace. Android itself also doesn't disable calls into its `Trace` API when in release mode.

This is already done for Fuchsia, and would be a small-ish change to do for Android as well. See:

https://github.com/dart-lang/sdk/blob/master/runtime/vm/globals.h#L111

and

https://github.com/dart-lang/sdk/blob/master/runtime/vm/timeline.cc#L6

**Comment by xster** [comment]
From @jacob314: we can offer an engine side hook to connect when the Dart APIs are called to record a Timeline event. Where we can branch out to the embedding to record it on the platform side.

**Comment by dnfield** [comment]
IOW: 

- If ATrace_isEnabled is true at shell creation, assume that the developer is profiling the application via systrace and use it.
- If the developer specifically requested trace_systrace, assume they will at some point profile using systrace and use it.
- Otherwise, use the VM timeline recorder.

**Comment by dnfield** [comment]
This still leaves a hole around things we could potentially be tracing before starting the VM, but that should be addressed separately as a Flutter specific bug.

**Comment by a-siva** [comment]
//cc @bkonyi
